server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -424,144 +410,215 @@ public ClusterState execute(ClusterState currentState) {
-                        rtBuilder.addAsRestore(updatedIndexMetadata, recoverySource);
-                        blocks.updateBlocks(updatedIndexMetadata);
-                        mdBuilder.put(updatedIndexMetadata, true);
-                        renamedIndex = updatedIndexMetadata.getIndex();
                     }
-                    for (int shard = 0; shard < snapshotIndexMetadata.getNumberOfShards(); shard++) {
-                        shardsBuilder.put(
-                            new ShardId(renamedIndex, shard),
-                            new RestoreInProgress.ShardRestoreStatus(clusterService.state().nodes().getLocalNodeId()));
+                    newIndexUUID = currentIndexMetadata.getIndexUUID();
+                    indexMdBuilder.settings(
+                        Settings.builder()
+                            .put(snapshotIndexMetadata.getSettings())
+                            .put(IndexMetadata.SETTING_INDEX_UUID, newIndexUUID));
+                    IndexMetadata updatedIndexMetadata = indexMdBuilder.index(renamedIndexName).build();
+                    rtBuilder.addAsRestore(updatedIndexMetadata, recoverySource);
+                    blocks.updateBlocks(updatedIndexMetadata);
+                    mdBuilder.put(updatedIndexMetadata, true);
+                    renamedIndex = updatedIndexMetadata.getIndex();
+                }
+                relationIndexUUIDs.compute(relationName, (_, indexUUIDs) -> {
+                    if (indexUUIDs == null) {
+                        indexUUIDs = new ArrayList<>();
+                    }
+                    indexUUIDs.add(newIndexUUID);
+                    return indexUUIDs;
+                });
+                for (int shard = 0; shard < snapshotIndexMetadata.getNumberOfShards(); shard++) {
+                    shardsBuilder.put(
+                        new ShardId(renamedIndex, shard),
+                        new RestoreInProgress.ShardRestoreStatus(clusterService.state().nodes().getLocalNodeId()));
+                }
+            }
+
+            for (var entry : relationIndexUUIDs.entrySet()) {
+                RelationName relationName = entry.getKey();
+                String newSchema = relationName.schema().replaceAll(
+                    request.schemaRenamePattern(),
+                    request.schemaRenameReplacement()
+                );
+                String newTableName = relationName.name().replaceAll(
+                    request.tableRenamePattern(),
+                    request.tableRenameReplacement()
+                );
+                RelationName newName = new RelationName(newSchema, newTableName);
+                List<String> indexUUIDs = entry.getValue();
+                RelationMetadata existingRelation = currentMetadata.getRelation(newName);
+                RelationMetadata snapshotRelation = snapshotMetadata.getRelation(relationName);
+                if (existingRelation instanceof RelationMetadata.Table existingTable) {
```

Comment:
Shouldn't we also support blob tables? Sorry, forgot to mention this initially.

Reply:
Currently we don't support blob snapshots at all (data is not contained in the indices).

The `resolveIndices` logic running earlier already filters out any blob indices if I'm not mistaken.

Reply:
Ah right, just verified that blob tables/indices aren't included inside the snapshot.